contents/javascript/javascript.md

```diff
@@ -0,0 +1,114 @@
+# Writing Testable Javascript
+
+JavaScript is a powerful language, and because it is powerful, there are multiple ways to go about doing the same thing. The end result is that multiple collaborators working on a single project can produce different code that does the same thing. That is why there is need for a standard to writing JavaScript code, especially on the front end.
+
+Good JavaScript code should be testable, composable and reusable.
+
+## Writing testable JavaScript
+
+### Avoid coupling with selectors
+When writing front end code we would encounter code that manipulates the document object model (DOM). Let's look at one such example.
+
+```js
+var DIV_STATUS_MESSAGE = '#statusMessagesToUser';
+
+function populateStatusMessageDiv(message, status) {
+    var $statusMessageDivToUser = $(DIV_STATUS_MESSAGE);
+    ...
+}
+```
+
+In this piece of code we see that the selector is intrinsically tied to the function. This means that the function is tightly tied with the selector and is not easily testable as the test code has to generate the same markup in the test suite for the function to hook up to. In order to prevent such tight coupling, it is advised to leave the selector as a parameter, or sometimes, simply pass in the element itself.
+
+```js
+function populateStatusMessage(selector, message, status) {
+    var $statusMessageDivToUser = $(selector);
+    ...
+}
+```
+
+### Split business logic and presentation code
+When we have to write code that generates markup, it requires business logic. However, mixing the two up is not a good idea. Let's look at the following example:
+
+```js
+$.ajax({
+    type: 'POST',
+    url: '/admin/adminStudentGoogleIdReset?' + params,
+    beforeSend: function() {
+        $(button).html("<img src='/images/ajax-loader.gif'/>");
+    },
+    error: function() {
+        $(button).html('An Error Occurred, Please Retry');
+    },
+    success: function(data) {
+        ...
+    }
+});
+```
+
+It order to test such a function, we would now have to has both logic and also the mock-up generated. Splitting the logic and markup in two separate functions will both make it easier to test and composable because now you can reuse code that generates the markup in multiple places. 
```

Comment:
`now have to has`

I guess this is a typo?

Also, `test and composable` doesn't sound quite right.